checker/src/call_graph.rs

```diff
@@ -0,0 +1,936 @@
+            |edge_idx, edge| {
+                self.graph.edge_endpoints(edge_idx).and_then(|endpoints| {
+                    if edges.contains(&endpoints) {
+                        None
+                    } else {
+                        edges.insert(endpoints);
+                        Some(edge.to_owned())
+                    }
+                })
+            },
+        );
+        self.update(graph)
+    }
+
+    /// Returns the set of nodes (Node indexes, which uniquely identify nodes)
+    /// That are reachable from the given start node.
+    ///
+    /// The underlying algorithm used to perform graph traveral is a BFS,
+    /// however, only one crate root is included from the traversal.
+    fn reachable_from(&self, start_node: NodeIdx) -> HashSet<NodeIdx> {
+        let mut reachable = HashSet::<NodeIdx>::new();
+        let mut bfs = Bfs::new(&self.graph, start_node);
+        let mut croot: Option<NodeIdx> = None;
+        while let Some(node_idx) = bfs.next(&self.graph) {
+            if let Some(node) = self.graph.node_weight(node_idx) {
+                if node.is_croot() {
+                    if croot.is_none() {
+                        croot = Some(node_idx);
+                        reachable.insert(node_idx);
+                    }
+                } else {
+                    reachable.insert(node_idx);
+                }
+            }
+        }
+        reachable
+    }
+
+    /// Filter out all nodes from the graph that are not reachable
+    /// via start node identifiable by `name`.
+    fn filter_reachable(&self, name: &str) -> CallGraph {
+        if let Some(start_node) = self.get_node_by_name(name) {
+            let reachable = self.reachable_from(start_node);
+            let graph = self.graph.filter_map(
+                |node_idx, node| {
+                    if reachable.contains(&node_idx) {
+                        Some(node.to_owned())
+                    } else {
+                        None
+                    }
+                },
+                |_, edge| Some(edge.to_owned()),
+            );
+            self.update(graph)
+        } else {
+            panic!(
+                "Failed to filter graph; could not find start node: {}",
+                name
+            );
+        }
+    }
+
+    /// Helper function for folding excluded nodes.
+    ///
+    /// Computes the set of reachable nodes reachable
+    /// via one or more excluded nodes.
+    fn reachable_raw_edges(
+        &self,
+        excluded: &MidpointExcludedMap,
+        initial_set: &HashSet<HalfRawEdge>,
+    ) -> HashSet<HalfRawEdge> {
+        let mut reachable = initial_set.clone();
+        let mut queue = VecDeque::<NodeIdx>::new();
+        for edge in reachable.iter() {
+            queue.push_back(edge.0);
+        }
+        while let Some(node_idx) = queue.pop_front() {
+            if let Some(entry) = excluded.get(&node_idx) {
+                for node in entry.1.iter() {
```

Comment:
entry.1 is a bit cryptic. Why not introduce local names (via a pattern inside the Some) to make things more obvious?

Reply:
Good idea, I'll update that